src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.Multiplexer.cs

```diff
@@ -0,0 +1,427 @@
+                        Buffer.BlockCopy(
+                            dataBuffer, dataOffset + bytesConsumed, // from
+                            dataBuffer, dataOffset, // to
+                            dataLength - bytesConsumed // for
+                        );
+#if DEBUG
+                        // for debugging purposes fill the removed data area with an easily
+                        // recognisable pattern so we can see if it is misused
+                        Span<byte> removed = dataBuffer.AsSpan(dataOffset + (dataLength - bytesConsumed), (dataOffset + bytesConsumed));
+                        removed.Fill(0xFF);
+#endif
+
+                        // then recreate the data span so that we're looking at the data
+                        // that has been moved
+                        data = dataBuffer.AsSpan(dataOffset, dataLength - bytesConsumed);
+                    }
+
+                    newDataLength = dataLength - bytesConsumed;
+                }
+            }
+
+            if (data.Length > 0 && !consumeRemainderPacket)
+            {
+                if (data.Length >= TdsEnums.HEADER_LEN)
+                {
+                    // we have enough bytes to read the packet header and see how
+                    // much data we are expecting it to contain
+                    int packetDataLength = Packet.GetDataLengthFromHeader(data);
+
+                    if (data.Length == TdsEnums.HEADER_LEN + packetDataLength)
+                    {
+                        if (!consumePartialPacket)
+                        {
+                            // we can tell the caller that they should directly consume the data in
+                            // the input buffer, this is the happy path
+                            consumeInputDirectly = true;
+                        }
+                        else
+                        {
+                            // we took some data from the input to reconstruct the partial packet
+                            // so we can't tell the caller to directly consume the packet in the
+                            // input buffer, we need to construct a new remainder packet and then
+                            // tell them to consume it
+                            remainderPacket = new Packet
+                            {
+                                Buffer = dataBuffer,
+                                DataLength = packetDataLength,
+                                CurrentLength = data.Length
+                            };
+                            consumeRemainderPacket = true;
+                            Debug.Assert(remainderPacket.HasHeader); // precondition of entering this block
+                            Debug.Assert(remainderPacket.HasDataLength); // must have been set at construction
+                            if (remainderPacket.CurrentLength >= remainderPacket.RequiredLength)
```

Comment:
Should this be `>` instead of ` >=` ?

Reply:
No, it's possible that the remainder packet we're producing could contain multiple packets of data so we take however much is left and then allow the next execution of this function to split it apart if it needs to. If we just took the data we expect given the current header then we might miss a following packet. If there is no following packet then it should be equal or the server has somehow sent us too much data without a packet.

But change it and give it a try. There are tests that should cover this just filter to multiplex to find them.